File: modules/benchmarks/arrow-scala-benchmarks/src/main/scala/zio/LeftBind.scala
package arrow.benchmarks.effects.scala.zio

import zio._

object LeftBind {


  def loop(depth: Int, size: Int, i: Int): UIO[Int] =
    if (i % depth == 0) IO.effectTotal[Int](i + 1).flatMap { loop(depth, size, _) }
    else if (i < size) loop(depth, size, i + 1).flatMap(i => IO.effectTotal(i))
    else IO.effectTotal(i)

  def unsafeIOLeftBindLoop(depth: Int, size: Int, i: Int): Int =
    ZIORTS.unsafeRun(loop(depth, size, i))

}

File: modules/benchmarks/arrow-kio-benchmarks/src/main/kotlin/kio/LeftBind.kt
package arrow.benchmarks.effects.kio

import it.msec.kio.UIO
import it.msec.kio.flatMap
import it.msec.kio.just
import it.msec.kio.runtime.Runtime.unsafeRunSyncAndGet

object LeftBind {

  fun loop(depth: Int, size: Int, i: Int): UIO<Int> =
    when {
        i % depth == 0 -> just(i + 1).flatMap { loop(depth, size, it) }
        i < size -> loop(depth, size, i + 1).flatMap { just(it) }
        else -> just(i)
    }

  fun leftBind(depth: Int, size: Int, i: Int) =
    unsafeRunSyncAndGet(loop(depth, size, i))
}
